### PR TITLE
[DON'T MERGE] HBase design proposal

### PR DESCRIPTION
**Pull request by xiaoyao1991**
For design review purpose. Still WIP for TBD items.

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

File: elide-hbase.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+      return rowKey;
+    }
+  }
+}
+```
+
+### Bridging DataStores
+1. Must use a `MultiplexDataStore`
+1. `PersistenceResource` will still be responsible for holding security check logics. `DataStoreTransaction` will have more to do with getting/setting attr, relations, etc. Methods in `DataStoreTransaction` aren't aware of securities at all
+1. Have a custom annotation `@DataStore(cls)`. This will be used to annotate relations of a different data store type in an entity bean.
```
Comment (aklish) [inline elide-hbase.md]: We'll need to know the relationship cardinality oneToOne, oneToMany, ManToOne, ManyToMany so the entity dictionary can reflect this.  I suggest we either make an elide set of these annotations OR have a single annotation with an attribute that describes this cardinality.

File: elide-hbase.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+1. Add a field `parentTransaction` to `DataStoreTransaction`, so that when using a `MultiplexTransaction`, a detail data store transaction object can get access to its parent(the MultiplexTransaction) and get to its peer(a transaction of a different data store type):
+```
+class DataStoreTransaction {
+  default DataStoreTransaction getParentTransaction() { return null; }
+}
+```
+1. Define `getRelation(Object entity, String relationName)` in `DataStoreTransaction`:
+```
+class DataStoreTransaction {
+  <T> Collection<T> getRelation(Object entity, String relationName) {
```
Comment (aklish) [inline elide-hbase.md]: We should also push filters, pagination, and sorting to the data store - so there should be a similar function with those args.

File: elide-hbase.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    // Default implementation should call the getter method(just like the getValue() in PersistenceResource)
+    // if the getter method is annotated with @DataStore, retrieve the transaction for the specified datastore from the parent multiplex transaction,
+    // and then use it to loadObjects, with the Criteria/FilterScope computed by @RowKeyFn(or maybe rename it) function.
+  }
+}
+```
+1. Define `updateRelation(Object entity, String fieldName, Set<PersistentResource> resourceIdentifiers)` in `DataStoreTransaction`:
```
Comment (aklish) [inline elide-hbase.md]: Do we need to distinguish between added and deleted here?

File: elide-hbase.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  boolean setRelation(Object entity, String fieldName, Set<PersistentResource> resourceIdentifiers) {
+    // Default implementation should call the setter method(just like the setValue() in PersistenceResource)
+    // if the setter method is annotated with @DataStore, retrieve the transaction for the specified datastore from the parent multiplex transaction,
+    // and then use it to save(serialize), with the Criteria/FilterScope computed by @RowKeyFn(or maybe rename it) function.
+  }
+}
+```
+
+#### Getting & Setting attributes
+Attributes of an entity should be mostly within the same data store. It's not usual that a MySQL entity has an *attribute* that lives in HBase(in those cases, a better data model design would be making that attribute a relation.)
```
Comment (aklish) [inline elide-hbase.md]: While I agree this is probably unusual, I'm not sure we can make the case that this won't be needed.

File: elide-hbase.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+### Bridging DataStores
+1. Must use a `MultiplexDataStore`
+1. `PersistenceResource` will still be responsible for holding security check logics. `DataStoreTransaction` will have more to do with getting/setting attr, relations, etc. Methods in `DataStoreTransaction` aren't aware of securities at all
+1. Have a custom annotation `@DataStore(cls)`. This will be used to annotate relations of a different data store type in an entity bean.
+```
+@DataStore(cls=HBaseStore.class)
+public Collection<HBaseEntity> getHBaseEntities() { return Collections.emptySet(); }
+```
+1. Add a field `parentTransaction` to `DataStoreTransaction`, so that when using a `MultiplexTransaction`, a detail data store transaction object can get access to its parent(the MultiplexTransaction) and get to its peer(a transaction of a different data store type):
```
Comment (aklish) [inline elide-hbase.md]: Why is this needed?

File: elide-hbase.md (hunk 1 above)
Comment (xiaoyao1991) [inline elide-hbase.md]: The `@DataStore` annotation is not for registering classes. It's used to annotate a relation(field or method) that any CRUD on that relation should use that dataStore specified in the annotation.

Comment (xiaoyao1991) [inline elide-hbase.md]: I think you're right that `@DataStore` is unnecessary.

File: elide-hbase.md (hunk 5 above)
Comment (xiaoyao1991) [inline elide-hbase.md]: My original thought is to have a parent transaction reference, so when seeing `@DataStore(cls)` annotation, we can get the transaction of the specific data store type by traversing up to the parent and down to the leaf. But after reading the comment above, `@DataStore` is not necessary, and transaction can be retrieved by looking up in `EntityDictionary` I guess.

File: elide-hbase.md (hunk 1 above)
Comment (xiaoyao1991) [inline elide-hbase.md]: Q: Will Hibernate (polyglot persistenc) respect the relationship annotations (OneToOne, etc) or will it blow up and think it needs to 'manage' the entity? 
A: Just experimented, Hibernate blows up when it sees a relationship annotated field which it is not aware of. Hibernate has its own validation logic on the validity of annotated fields. To workaround that may involve overriding and replacing a lot of Hibernate code. So it looks like an easy way around is to leverage the `@Transient` and `@ComputedAttributes` to annotate the relation fields.

Comment (xiaoyao1991) [inline elide-hbase.md]: One thing that may be semantically incorrect if we do that way is that, with `@ComputedAttribute`, the results of that field will appear in the `attributes` section of the JSON output. Ideally, I would prefer to have a `@ComputedRelation` where the results of that field appear in the `relationships` section of the JSON

File: elide-hbase.md (hunk 4 above)
Comment (xiaoyao1991) [inline elide-hbase.md]: I feel like attributes are mostly just primitives or simple object types(like Date). I think it'll be hard to know which dataStore to source from for a given primitive attribute.
